Clear user service cache on transaction end

chdorner opened the pull request:
We previously cleared the user service cache on `after_commit` and
`after_rollback` SQLAlchemy events. This caused issues where the view
was raising an exception, because SQLAlchemy only fires the
`after_rollback` event for an explicit rollback.

pyramid_tm and zope.sqlalchemy will close the database transaction when
the view raises an exception. A call to `SessionTransaction.close` does
implicitly send a `ROLLBACK` SQL query to the database, but it doesn't
emit the `after_rollback` event that we're relying on in order to clear
out the user service cache.
This results in a exception views potentially accessing detached
database objects which in turn throws another exception, which then
returns a generic 500 error back to the user.

For following all of this along at home:

1. pyramid_tm starts a transaction at the beginning of a request:
https://github.com/Pylons/pyramid_tm/blob/1.1.1/pyramid_tm/__init__.py#L86
2. pyramid_tm catches exceptions and if the error is not retryable,
aborts the transaction:
https://github.com/Pylons/pyramid_tm/blob/1.1.1/pyramid_tm/__init__.py#L116
3. The `abort` that pyramid_tm sends is finishing the transaction:
https://github.com/zopefoundation/zope.sqlalchemy/blob/0.7.7/src/zope/sqlalchemy/datamanager.py#L91
4. Finishing a transaction in turn will close it:
https://github.com/zopefoundation/zope.sqlalchemy/blob/0.7.7/src/zope/sqlalchemy/datamanager.py#L87

Whereas an explicit rollback works as we expected it to:
1. `sqlalchemy.orm.session.SessionTransaction.rollback` checks if it
needs to send a `ROLLBACK` SQL query to the database, and if it does...
(https://bitbucket.org/zzzeek/sqlalchemy/src/ce17fd4bf361e6cd08b6e698e49c118643abe724/lib/sqlalchemy/orm/session.py?at=rel_1_1_4&fileviewer=file-view-default#session.py-490)
2. calls `_rollback_impl` which emits the `after_rollback` event
(https://bitbucket.org/zzzeek/sqlalchemy/src/ce17fd4bf361e6cd08b6e698e49c118643abe724/lib/sqlalchemy/orm/session.py?at=rel_1_1_4&fileviewer=file-view-default#session.py-532).

Fixes #4398.

nickstenning commented on h/services/user.py:
```diff
@@ -40,10 +40,18 @@ def __init__(self, default_authority, session):
+        @sqlalchemy.event.listens_for(session, 'after_transaction_end')
+        def flush_cache(session, transaction):
+            if transaction.parent is None:
+                # We only clear the cache when the top-level transaction
```
This comment would make more sense to me above the conditional.

nickstenning commented:
BTW, if it makes testing easier, you might want to pull out the "after_transaction_end" stuff into a utility decorator in `h/utils/something.py`, and then use that decorator in the user service and groupfinder service. That way you can just test that the services use the right decorator, and you only need to do the crazy stuff with the `db_session` fixture once.

nickstenning commented:
And, frankly, you might not need to do it at all, because at that point you're really going to be testing SQLAlchemy...

chdorner commented:
I've refactored this into a util function, which is tested by emitting the appropriate event in the tests which allows us to only test the implementation and not SQLAlchemy. In addition to that, I changed the groupfinder service as well.

This should be ready to review again.

nickstenning commented on h/util/db.py:
```diff
@@ -0,0 +1,36 @@
+    code after the view, exception views for example, will not be able to access
+    detached instances.
+
+    The wrapped function will receive the database session as ``session`` and the
```
Is there any reason to pass these through at the moment? I think I'd lean towards not doing so unless there's a need, as it will at least prevent unused argument warnings.

nickstenning commented on h/util/db.py:
```diff
@@ -0,0 +1,36 @@
+           self._cache = {}
+
+    """
+    def wrapped(func):
```
The naming here is a bit confusing, because `wrapped` is actually a decorator, not a wrapped function. Calling this `decorate` would probably be a bit clearer, and is [sort of conventional](https://github.com/zzzeek/sqlalchemy/blob/c04870ba7b8098c7d408ad66f60efe7229496fde/lib/sqlalchemy/event/api.py#L92-L126).

nickstenning commented on h/util/db.py:
```diff
@@ -0,0 +1,36 @@
+        def after_transaction_end(current_session, transaction):
+            # We only clear the cache when the top-level transaction finishes.
+            if transaction.parent is None:
+                func(current_session, transaction)
```
So this is a decorator, which means it should return the function that it's wrapping. In this particular case that's easy, because you don't actually need to modify the function at all (that's not we made it a decorator), so you can use [a similar pattern to the `listens_for` decorator itself](https://github.com/zzzeek/sqlalchemy/blob/c04870ba7b8098c7d408ad66f60efe7229496fde/lib/sqlalchemy/event/api.py#L92-L126).

So maybe something like:

```python
def on_transaction_end(session):
    def decorate(func):
        def _handler(_, transaction):
            if transaction.parent is None:
                func()
        sqlalchemy.event.listen(session, 'after_transaction_end', _handler)
        return func
    return decorate
```

nickstenning commented on h/util/db.py:
```diff
@@ -0,0 +1,35 @@
+    """
+    Decorator for a function which should run after a top-level transaction ended.
+
+    Transaction that are either implicitly or explicity committed or rolled back will be
```
*Transactions

nickstenning replied:
*explicitly
